cloudinit/sources/helpers/vultr.py

```diff
@@ -20,26 +22,65 @@ def get_metadata(url, timeout, retries, sec_between, agent):
     exception = RuntimeError("Failed to DHCP")
 
     # Seek iface with DHCP
-    for iface in net.get_interfaces():
-        # Skip dummy, lo interfaces
-        if "dummy" in iface[0]:
-            continue
-        if "lo" == iface[0]:
-            continue
+    for iface in get_interface_list():
         try:
             with EphemeralDHCPv4(
                 iface=iface[0], connectivity_url_data={"url": url}
             ):
+                # Check for the metadata route, skip if not there
+                if not check_route():
+                    continue
+
                 # Fetch the metadata
                 v1 = read_metadata(url, timeout, retries, sec_between, agent)
 
                 return json.loads(v1)
-        except (NoDHCPLeaseError, subp.ProcessExecutionError) as exc:
+        except (
+            NoDHCPLeaseError,
+            subp.ProcessExecutionError,
+            RuntimeError,
+            exceptions.RequestException,
+        ) as exc:
             LOG.error("DHCP Exception: %s", exc)
             exception = exc
     raise exception
 
 
+# Get interface list, sort, and clean
+# Should be replaced with find_candidate_nics
+# Dummy and lo will remain to need to be removed
+def get_interface_list():
+    ifaces = []
+    for iface in net.get_interfaces():
+        # Skip dummy, lo interfaces
+        if "dummy" in iface[0]:
+            continue
+        if "lo" == iface[0]:
+            continue
+        ifaces.append(iface)
+
+    # Sort alphabetically
+    return sorted(ifaces, key=lambda i: i[0])
+
+
+# Check for /32 route that our dhcp servers inject
+# in order to determine if this a customer-run dhcp server
+def check_route():
+    # Get routes, confirm entry exists
+    routes = netinfo.route_info()
+
+    # If no tools exist and empty dict is returned
+    if "ipv4" not in routes:
+        return False
+
+    # Parse each route into a more searchable format
+    for route in routes["ipv4"]:
+        if "169.254.169.254" in route["destination"]:
```

Comment:
Is a check needed here? Can route ever not have destination?

Reply:
```
if route.get("destination", None) == "169.254.159.254":
    return True
```

Might be more nilsafe and remain a one-liner.

Reply:
I'm also curious as to whether we should be explicitly referring to the IP string literal here. Referencing `BULTIN_DNS_CONFIG["url"]` might be easier to maintain in the long run and more meaningful.

Reply:
> I'm also curious as to whether we should be explicitly referring to the IP string literal here. Referencing `BULTIN_DNS_CONFIG["url"]` might be easier to maintain in the long run and more meaningful.

Could be useful if someone does something weird here or if we change this (doubtful). However I adjusted to see if the route is in the url to avoid having to do any wacky regex or parsing.